### PR TITLE
snapstate: use snapstate.Type in backend.RemoveSnapFiles

### PR DESCRIPTION
**Pull request by mvo5**
This branch makes use of the fact that we have the type of the snap in the snapstate now.

### CODE REVIEW

File: overlord/snapstate/backend_test.go [hunk 1]
```diff
@@ -216,7 +216,7 @@ func (f *fakeSnappyBackend) UnlinkSnap(info *snap.Info, meter progress.Meter) er
 	return nil
 }
 
-func (f *fakeSnappyBackend) RemoveSnapFiles(s snap.PlaceInfo, meter progress.Meter) error {
+func (f *fakeSnappyBackend) RemoveSnapFiles(s snap.PlaceInfo, typ snap.Type, meter progress.Meter) error {
```
Comment (niemeyer) [inline overlord/snapstate/backend_test.go]: Should we put those into the fakeOp to ensure we're handing in the right data?

File: overlord/snapstate/backend/setup.go [hunk 2]
```diff
@@ -77,15 +76,10 @@ func (b Backend) RemoveSnapFiles(s snap.PlaceInfo, meter progress.Meter) error {
...
-		info, _, err := OpenSnapFile(snapPath, nil)
-		if err != nil {
-			return fmt.Errorf("cannot read snap file %q to determine snap type: %v", snapPath, err)
-		}
-
+	// snapPath may either be a file or a (broken) symlink to a dir
+	if _, err := os.Lstat(snapPath); err == nil {
 		// remove the kernel assets (if any)
-		if info.Type == snap.TypeKernel {
+		if typ == snap.TypeKernel {
```
Comment (niemeyer) [inline overlord/snapstate/backend/setup.go]: This is probably not tested much, but the situation is not being made any worse, so sounds fine.

**Comment by niemeyer** [comment]
LGTM

File: overlord/snapstate/backend/setup.go (hunk 2 above)
Comment (mvo5) [inline overlord/snapstate/backend/setup.go]: We have (indirect) tests in setup_test.go for this.

File: overlord/snapstate/snapmgr.go [hunk 3]
```diff
@@ -641,14 +645,18 @@ func snapSetupAndState(t *state.Task) (*SnapSetup, *SnapState, error) {
...
-	ss, _, err := snapSetupAndState(t)
+	ss, snapst, err := snapSetupAndState(t)
 	t.State().Unlock()
 	if err != nil {
 		return err
 	}
 
 	pb := &TaskProgressAdapter{task: t}
-	return m.backend.UndoSetupSnap(ss.placeInfo(), pb)
+	typ, err := snapst.Type()
```
Comment (pedronis) [inline overlord/snapstate/snapmgr.go]: is this case fully tested?  atm the code is setting type very late (doLinkSnap) so I think in some scenarios we will tend to explode there (we need a test failing in CopyData I think),  

we can try to set Type earlier, it was set late to cope with preexisting changes
safest maybe to set it both early and late for a while

Comment (pedronis) [inline overlord/snapstate/snapmgr.go]: to be clearer what I'm saying is:
* we need  a test
* we likely want to set snapst.SeType() in doMountSnap as well

Comment (mvo5) [inline overlord/snapstate/snapmgr.go]: Thanks, good catch! I added a test and this showed that indeed the type needs to be set earlier. This is done now as well.

File: overlord/snapstate/backend_test.go (hunk 1 above)
Comment (mvo5) [inline overlord/snapstate/backend_test.go]: Yes, I added this.

**Pull request merged**